colexec: drain Materializer metadata sources as inputs in ProcessorBase

asubiotto opened the pull request:
Some errors, like ReadWithinUncertaintyInterval errors, should be swallowed if
all results have been returned already.

Release note: None (fix for bug not in release)

Fixes #50299

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50388)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
I think this approach is essentially the same as in the last version of my PR with the difference being that here when in `StateTrailingMeta` we swallow errors only in the parallel unordered synchronizer, and there we do so in all components. If these are the two options, I think my approach makes more sense. The build failures are also the same on both PRs - context cancelation error propagated first.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-:-MA6lenv6UKpPA5XO-CE:btsw8q)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto commented:
Urgh, didn't notice that failure. I think the difference is that in your PR you're swallowing errors in `StateTrailingMeta`, while here, we're adding swallowing to the equivalent of `StateDraining` in the `ParallelUnorderedSynchronizer`. We shouldn't be swallowing any errors in `StateTrailingMeta` because no errors should be generated (this is after forwarding meta) so swallowing in that stage is redundant. I'll take a look at the failure.

asubiotto commented:
Hmm, I think that a problem is that the `Materializer` is draining its inputs in the trailing meta state. The correct thing would be to wrap the `metadataQueue` in `RowSources` and drain them in `StateDraining`. This would guarantee correct semantics for all vectorized components. This is not related to the context cancellation.

yuzefovich commented:
> because no errors should be generated (this is after forwarding meta) so swallowing in that stage is redundant

I think this is not true currently in the vectorized components. For example, in `Columnarizer.Next`  we simply buffer any metadata we get without checking whether it contains an error and return all of it in `DrainMeta` (which occurs in `StateTrailingMeta`). Similarly, in `Inbox` we buffer all metadata received from the stream as the message, without checking whether it contains an error. These two places seem to be the only ones where we have such behavior.

asubiotto commented:
Yeah, my second comment was regarding that. We should move the draining of vectorized components to `StateDraining` in the materializer, which will delegate to `ProcessorBase` to do this filtering for us.
By the way, the context cancellation is occurring because the parallel unordered synchronizer cancels inputs when it receives the uncertainty error, and this is being returned as metadata.

asubiotto commented:
Updated with the materializer change and fix for context cancellation.

yuzefovich reviewed (commented):
> Reviewed 3 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7AozW1EnjS903hc5w:-MA7AozW1EnjS903hc5x:btcy8mi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L74)):*
> ```Go
> // interface. This is done so that the Materializer can drain MetadataSources
> // in the vectorized input tree as inputs, rather than draining them in the
> // trailing metadata state, which should is meant only for internal metadata
> ```

nit: "should is meant".
___
*[pkg/sql/colexec/materializer.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7AuytEzzoJl07Hl6x:-MA7AuytEzzoJl07Hl6y:b2zx3oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L92)):*
> ```Go
> // OutputTypes implements the RowSource interface.
> func (d *drainHelper) OutputTypes() []*types.T {
> 	panic("unimplemented")
> ```

The linter will complain, use `colexecerror.InternalError`.
___
*[pkg/sql/colexec/materializer.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7B80G6FbgQz73W-M4:-MA7B80G6FbgQz73W-M5:b-iyasse) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L104)):*
> ```Go
> func (d *drainHelper) Next() (sqlbase.EncDatumRow, *execinfrapb.ProducerMetadata) {
> 	if d.bufferedMeta == nil {
> 		d.bufferedMeta = d.DrainMeta(d.ctx)
> ```

I think `DrainMeta` is free to return `nil`, so maybe in order to avoid multiple calls to `DrainMeta` we could add this check
```
if d.bufferedMeta == nil {
  d.bufferedMeta = []execinfrapb.ProducerMetadata{}
}
```
___
*[pkg/sql/colexec/materializer.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7Beqt7XovBDevB8_3:-MA7Beqt7XovBDevB8_4:bhvfiqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L165)):*
> ```Go
> 		nil, /* memMonitor */
> 		execinfra.ProcStateOpts{
> 			InputsToDrain: []execinfra.RowSource{m.drainHelper},
> ```

I think this change is not entirely correct, philosophically speaking. For example, wrapped processors (like `indexJoiner`) as well as `colBatchScan` actually do generate metadata that is propagated in `StateTrailingMeta` in the row-based flows. I'm not sure how important to fix it though.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/materializer.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7AuytEzzoJl07Hl6x:-MAQTq3E5Raa5v8RBf5o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L92)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

The linter will complain, use `colexecerror.InternalError`.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/materializer.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7B80G6FbgQz73W-M4:-MAQU4yz8yxP1d_JPGA1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L104)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think `DrainMeta` is free to return `nil`, so maybe in order to avoid multiple calls to `DrainMeta` we could add this check
```
if d.bufferedMeta == nil {
  d.bufferedMeta = []execinfrapb.ProducerMetadata{}
}
```
</blockquote></details>

Done.
___
*[pkg/sql/colexec/materializer.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7Beqt7XovBDevB8_3:-MAQU5dk2tIh3YV1uhiE:b-fbpaxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L165)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this change is not entirely correct, philosophically speaking. For example, wrapped processors (like `indexJoiner`) as well as `colBatchScan` actually do generate metadata that is propagated in `StateTrailingMeta` in the row-based flows. I'm not sure how important to fix it though.
</blockquote></details>

From the perspective of a downstream processor, metadata generated in the trailing meta state of upstream processors is always received in `StateDraining` of the processor. In other words, it does not matter when an upstream processor generated a bit of metadata, all metadata drained from inputs is treated equally. The difference is that **forwarding** metadata needs to happen in `StateDraining` but **generating** metadata needs to happen in `StateTrailingMeta`. Does that make sense?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7Beqt7XovBDevB8_3:-MARcUsA5t_031RHAQ-0:bmfhjlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L165)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

From the perspective of a downstream processor, metadata generated in the trailing meta state of upstream processors is always received in `StateDraining` of the processor. In other words, it does not matter when an upstream processor generated a bit of metadata, all metadata drained from inputs is treated equally. The difference is that **forwarding** metadata needs to happen in `StateDraining` but **generating** metadata needs to happen in `StateTrailingMeta`. Does that make sense?
</blockquote></details>

Hm, I see. I think there was a misunderstanding on my part when I introduced `MetadataSource` interface. Currently, implementations of `DrainMeta` can generate metadata (e.g. `indexJoiner`), and this `drainHelper` will call `DrainMeta` when the index joiner is in `StateDraining`, so the metadata generation occurs in the wrong state. It seems like we need two different methods - `DrainMeta` - for forwarding the metadata that was generated while a processor was running - and `GenerateMeta` - for the generating trailing meta. Thoughts?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/materializer.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7Beqt7XovBDevB8_3:-MAWU7sN8QfAwQnYpyNB:b-q29p3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L165)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, I see. I think there was a misunderstanding on my part when I introduced `MetadataSource` interface. Currently, implementations of `DrainMeta` can generate metadata (e.g. `indexJoiner`), and this `drainHelper` will call `DrainMeta` when the index joiner is in `StateDraining`, so the metadata generation occurs in the wrong state. It seems like we need two different methods - `DrainMeta` - for forwarding the metadata that was generated while a processor was running - and `GenerateMeta` - for the generating trailing meta. Thoughts?
</blockquote></details>

I think that's fine. The important part is to avoid forwarding metadata in the trailing meta state. It doesn't seem like there's much point in avoiding generating metadata in `StateDraining` (my first comment was probably unnecessarily restrective) . In an ideal world we'd probably want these two states to be one but it does allow us to prioritize metadata (forward first, then generate) which makes sense.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-:-MAWZFLS-yQsn1oWrReX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/colexec/materializer.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50388#-MA7Beqt7XovBDevB8_3:-MAWZB-B9OoPF9DIZPcX:b2nhll9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c12379289d6653a2f932dbf1ae5c7827be5d43f1/pkg/sql/colexec/materializer.go#L165)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think that's fine. The important part is to avoid forwarding metadata in the trailing meta state. It doesn't seem like there's much point in avoiding generating metadata in `StateDraining` (my first comment was probably unnecessarily restrective) . In an ideal world we'd probably want these two states to be one but it does allow us to prioritize metadata (forward first, then generate) which makes sense.
</blockquote></details>

Ok.


<!-- Sent from Reviewable.io -->

Commit:
colexec: don't forward internal cancellation error in parallel synchronizer

The ParallelUnorderedSynchronizer terminates spawned goroutines using context
cancellation when encountering an error. Previously, these goroutines would
forward a cancellation error indiscriminately. The behavior remains unchanged
in case of cancellation by external components.

Release note: None (behavior that is fixed not present in a release)

Commit:
colexec: drain Materializer metadata sources as inputs in ProcessorBase

The materializer was previously draining its inputs as part of producing
trailing metadata. According to ProcessorBase, this state is meant only for
generating internal metadata. StateDraining is the state before this that
forwards any metadata generated from inputs. In order to respect this contract,
the Materializer now wraps metadata sources in a helper and passes these as
inputs to drain in StateDraining to the ProcessorBase to handle metadata
forwarding correctly.

Release note: None (fix for bug not in release)

asubiotto commented:
Fixed the test issue, which was that the internal cancellation would not affect the `ctx` passed in to `Next` in an input chain, so that could stay hung despite internal cancellation. Went down the approach of setting an atomic if the cancellation is internal.

asubiotto commented:
bors r=yuzefovich
